zookeeper-server/src/main/java/org/apache/zookeeper/server/SyncRequestProcessor.java

```diff
@@ -178,27 +226,22 @@ public void run() {
                 ServerMetrics.getMetrics().SYNC_PROCESSOR_QUEUE_TIME.add(startProcessTime - si.syncQueueStartTime);
 
                 // track the number of records written to the log
-                if (!si.isThrottled() && zks.getZKDatabase().append(si)) {
-                    if (shouldSnapshot()) {
-                        resetSnapshotStats();
+                if (!si.isThrottled() && zkDB.append(si)) {
+                    if (selfSnapTxnRollThreshold.meet(zkDB.getTxnCount(), zkDB.getTxnSize())) {
                         // roll the log
-                        zks.getZKDatabase().rollLog();
-                        // take a snapshot
-                        if (!snapThreadMutex.tryAcquire()) {
-                            LOG.warn("Too busy to snap, skipping");
+                        zkDB.rollLog();
+
+                        if (onlySnapWhenSafetyIsThreatened) {
+                            if (safetySnapThreshold.meet(zkDB.getTxnsSinceLastSnap(), zkDB.getTxnsSizeSinceLastSnap())) {
+                                snapGenerator.takeSnapshot(false);
+                                ServerMetrics.getMetrics().TAKING_SAFE_SNAPSHOT.add(1);
+                                LOG.info("The txns size since last snapshot is larger than 10x, "
+                                        + "going to take a snapshot for safe.");
+                            }
                         } else {
-                            new ZooKeeperThread("Snapshot Thread") {
-                                public void run() {
-                                    try {
-                                        zks.takeSnapshot();
-                                    } catch (Exception e) {
-                                        LOG.warn("Unexpected exception", e);
-                                    } finally {
-                                        snapThreadMutex.release();
-                                    }
-                                }
-                            }.start();
+                            snapGenerator.takeSnapshot(false);
```

Comment:
similar here - should the parameter be hardcoded or taken from fsyncSnapshotFromScheduler